### DISCUSSION TITLE
[java] Introduce a new rule package-by-feature?

### DISCUSSION THREAD

**Discussion by linusjf** [pmd/pmd#3023] [Q&A] [answered]
**Description:**

Is there scope for a rule to limit the number of classes in a package? This can be parameterized by whether to include inner classes or not. The default is not to. Test classes, if present in the same package, must be excluded.

https://dzone.com/articles/rule-30-%E2%80%93-when-method-class-or

https://stackoverflow.com/questions/312642/how-many-classes-per-package-methods-per-class-lines-per-method/312650

"A package shouldn’t contain more than 30 classes, thus comprising up to 27,000 code lines.

Subsystems with more than 30 packages should be avoided. Such a subsystem would count up to 900 classes with up to 810,000 lines of code.

 A system with 30 subsystems would thus possess 27,000 classes and 24.3 million code lines."

Another related design rule could be whether to flag a package as not following package-by-feature.

http://www.javapractices.com/topic/TopicAction.do?Id=205

"...in order for package-by-feature to really work as desired, all items related to a given feature - from user interface, to Java code, to database items - must be placed in a single directory dedicated to that feature (and only that feature).

In some cases, a feature/package will not be used by any other feature in the application. If that's the case, it may be removed simply by deleting the directory. If it is indeed used by some other feature, then its removal will not be as simple as a single delete operation.

That is, the package-by-feature idea does not imply that one package can never use items belonging to other packages. Rather, package-by-feature aggressively prefers package-private as the default scope, and only increases the scope of an item to public only when needed."

This, of course, leads to an interesting question as how to unit test package-private classes.

Should we then exclude test classes in the same package from the classes count for that package?

https://stackoverflow.com/questions/42255972/how-can-we-test-package-private-class/42256510#42256510

See answer above.

**Comment by linusjf** [comment]
> Another related design rule could be whether to flag a package as not following package-by-feature.
> 
> http://www.javapractices.com/topic/TopicAction.do?Id=205
> 
> "...in order for package-by-feature to really work as desired, all items related to a given feature - from user interface, to Java code, to database items - must be placed in a single directory dedicated to that feature (and only that feature).
> 
> In some cases, a feature/package will not be used by any other feature in the application. If that's the case, it may be removed simply by deleting the directory. If it is indeed used by some other feature, then its removal will not be as simple as a single delete operation.
> 
> That is, the package-by-feature idea does not imply that one package can never use items belonging to other packages. Rather, package-by-feature aggressively prefers package-private as the default scope, and only increases the scope of an item to public only when needed."
> 

Package-by-feature must also include sub-packages under the same directory structure. This will also allow unit tests that have the tests in a separate sub-package *tests*. This retains the delete directory property deleting the feature. The package classes can refer to other packages but care must be taken to ensure that packages that refer classes in this package mark or document it as a dependency; the dependency has to be one-way, not bi-directional.

**Comment by linusjf** [comment]
https://phauer.com/2020/package-by-feature/

https://hackernoon.com/package-by-features-not-layers-2d076df1964d

https://dzone.com/articles/package-by-feature-is-demanded

I'm not quite sure yet as to how this rule can be enforced. Or whether the existing coupling rules suffice.

Perhaps, the [LoosePackageCoupling](https://pmd.github.io/latest/pmd_rules_java_design.html#loosepackagecoupling) rule can be cannibalised to specify packages and classes by package name. Each package-by-feature can be configured separately. 

You could simply restrict the API signatures to packages and sub-packages within the top package but that might be too restrictive.

**Comment by adangel** [comment]
I'm not sure, whether you are asking a question (and which one) or whether you are suggesting to add  a new rule...

If you want to know, whether PMD already has a rule to limit the number of classes in a package, then, I'd suggest to look up the rule reference -> https://pmd.github.io/latest/pmd_rules_java.html
I'd search such a rule in the "Design" category.

So far, I don't see a match. Similar rules are [ExcessiveClassLength](https://pmd.github.io/latest/pmd_rules_java_design.html#excessiveclasslength), [ExcessiveMethodLength](https://pmd.github.io/latest/pmd_rules_java_design.html#excessivemethodlength), [ExcessivePublicCount](https://pmd.github.io/latest/pmd_rules_java_design.html#excessivepubliccount), [TooManyMethod](https://pmd.github.io/latest/pmd_rules_java_design.html#toomanymethods) - but they all address issues on a finer level. 

So, answer is: No, there is no such rule yet.

For the package-by-feature structure, [LoosePackageCoupling](https://pmd.github.io/latest/pmd_rules_java_design.html#loosepackagecoupling) might help, but I'm not sure, whether it's flexible enough already.

**Comment by linusjf** [comment, marked as answer]
> I'm not sure, whether you are asking a question (and which one) or whether you are suggesting to add a new rule...
> 
> If you want to know, whether PMD already has a rule to limit the number of classes in a package, then, I'd suggest to look up the rule reference -> https://pmd.github.io/latest/pmd_rules_java.html
> I'd search such a rule in the "Design" category.
> 
> So far, I don't see a match. Similar rules are [ExcessiveClassLength](https://pmd.github.io/latest/pmd_rules_java_design.html#excessiveclasslength), [ExcessiveMethodLength](https://pmd.github.io/latest/pmd_rules_java_design.html#excessivemethodlength), [ExcessivePublicCount](https://pmd.github.io/latest/pmd_rules_java_design.html#excessivepubliccount), [TooManyMethod](https://pmd.github.io/latest/pmd_rules_java_design.html#toomanymethods) - but they all address issues on a finer level.
> 
> So, answer is: No, there is no such rule yet.



Yes, so I've written up a [separate issue](https://github.com/pmd/pmd/issues/2935) after having investigated further and after arriving at the conclusion that it's doable.



> 
> For the package-by-feature structure, [LoosePackageCoupling](https://pmd.github.io/latest/pmd_rules_java_design.html#loosepackagecoupling) might help, but I'm not sure, whether it's flexible enough already.

I don't think it is. This would have to be specified for each package implementing a feature. 

I haven't used this rule myself. It appears to be controversial and probably least used, since it has to be configured by the user diligently.
